apps/music/js/ui/views/player_view.js

```diff
@@ -540,63 +393,57 @@ var PlayerView = {
-            this.pause();
-          }
-        }.bind(this)).catch(function(msg) {
-          console.error(msg);
-        });
-      }.bind(this));
-    } else if (this.sourceType === TYPE_BLOB && !this.audio.src) {
-      // When we have to play a blob, we need to parse the metadata
-      this.getMetadata(this.dataSource, function(metadata) {
-        // Add the blob from the dataSource to the fileinfo because we want use
-        // the cover image which embedded in that blob so that we don't have to
-        // count on the database.
-        this.setInfo({metadata: metadata,
-                      name: this.dataSource.name,
-                      blob: this.dataSource});
-
-        this.setAudioSrc(this.dataSource);
-      }.bind(this));
-    } else {
-      // If we reach here, the player is paused so resume it.
-      // But if we're very close to the end of the song (and if the song
-      // is not really short) then just skip to the next song rather than
-      // finishing this one. (This works around Bug 1157118 where if we're
-      // within a fraction of a second of the end of a .m4a file, it takes
-      // a long time to get an ended event and move to the next song.)
-      if (this.audio.duration > 20 &&
-          this.audio.duration - this.audio.currentTime < 1) {
-        this.next(true);
+    this.queue.current().then((songData) => {
+      this.setInfo(songData);
+
+      if (songData.blob) {
+        return songData.blob;
       }
-      else {
-        this.audio.play();
+
+      this.setRatings(songData.metadata.rated);
+      Database.incrementPlayCount(songData);
+      return Database.getFile(songData, true);
+    }).then((file) => {
+      this.setAudioSrc(file);
+
+      // When we need to preview an audio like in picker mode,
+      // we will not autoplay the picked song unless the user taps to play
+      // And we just call pause right after play.
+      // Also we pause at beginning when SCO is enabled, the user can still
+      // select songs to the player but it won't start, they have to wait
+      // until the SCO is disconnected.
+      if (this.type === TYPE_SINGLE || MusicComms.isSCOEnabled) {
+        this.pause();
       }
+    }).catch(function(msg) {
+      console.error(msg);
+    });
+  },
+
+  play: function pv_play(targetIndex) {
+    this.checkSCOStatus();
+    this._sendInterpageMessage();
+    this.showInfo();
+
+    // The player must be paused, so let's resume it. However, if we're very
+    // close to the end of the song (and if the song is not really short) then
+    // just skip to the next song rather than finishing this one. (This works
+    // around Bug 1157118 where if we're within a fraction of a second of the
+    // end of a .m4a file, it takes a long time to get an ended event and move
+    // to the next song.)
+    if (this.audio.duration > 20 &&
+        this.audio.duration - this.audio.currentTime < 1) {
```

Comment:
This is an interesting hack! :-)

I wonder if we should just look at whether or not the file is an .m4a file too.